renderdoc/os/win32/win32_shellext.cpp

```diff
@@ -362,18 +399,127 @@ struct RDCThumbnailProvider : public IThumbnailProvider, IInitializeWithStream
       return E_NOTIMPL;
     }
 
-    const byte *jpgbuf = m_Thumb.pixels;
     size_t thumblen = m_Thumb.len;
     uint32_t thumbwidth = m_Thumb.width, thumbheight = m_Thumb.height;
+    byte *thumbpixels = NULL;
 
-    if(jpgbuf == NULL)
-      return E_NOTIMPL;
+    if(m_Thumb.format == FileType::JPG)
+    {
+      const byte *jpgbuf = m_Thumb.pixels;
+      if(jpgbuf == NULL)
+        return E_NOTIMPL;
+
+      int w = thumbwidth;
+      int h = thumbheight;
+      int comp = 3;
+      thumbpixels = jpgd::decompress_jpeg_image_from_memory(jpgbuf, (int)thumblen, &w, &h, &comp, 3);
+    }
+    else
+    {
+      const ResourceFormatType resourceType = m_ddsData.format.type;
+      const uint32_t decompressedStride = thumbwidth * 4;
+      const uint32_t decompressedBlockWidth = 16;    // in bytes (4 byte/pixel)
+      const uint32_t decompressedBlockMaxSize = 64;
+      const uint32_t compressedBlockSize = m_ddsData.format.ElementSize();
+
+      thumbwidth = m_ddsData.width;
+      thumbheight = m_ddsData.height;
+      // Delete buffers
+      delete[] m_ddsData.subdata;
+      delete[] m_ddsData.subsizes;
```

Comment:
The deletion of `m_ddsData.subdata`and `m_ddsData.subsizes` is now moved to line 160. I did lookup how ddsData handles the creation of the arrays and tried to its content based on that with a loop.

Reply:
This looks good now but there's one more minor issue here - the `m_Thumb.pixels` copy you make in lines 150-160 is allocated with `malloc`, but it should be `new[]` to match the other management of that pointer.

It's a bit annoying that there's this inconsistency, normally I don't use `malloc` in my code so it would be consistent. It's needed specifically only for `thumbpixels` since it should be compatible with the other path that uses the jpeg decoder, and that uses `malloc`. That should be the only case it's needed.

Reply:
Replaced to `buf = new byte[m_Thumb.len]`.